conn_browser.go

```diff
@@ -0,0 +1,229 @@
+			// connection to be closed.
+			select {
+			case <-c.dataSignal:
+				continue
+			case <-c.closeSignal:
+				return 0, io.EOF
+			}
+		} else {
+			return n, err
+		}
+	}
+}
+
+// checkOpen returns an error if the connection is not open. Otherwise, it
+// returns nil.
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errConnectionClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	uint8Array := js.Global().Get("Uint8Array").New(len(b))
+	for i := 0; i < len(b); i++ {
+		uint8Array.SetIndex(i, b[i])
+	}
+	c.Call("send", uint8Array.Get("buffer"))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) setUpHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			// TODO(albrow): Currently we assume data is of type Blob. Really we
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			data := readBlob(blob)
+			c.incomingData <- data
+		}()
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(c.closeSignal)
+		return nil
+	})
+	c.closeHandler = &closeHandler
+	c.Call("addEventListener", "close", closeHandler)
+}
+
+// readLoop continuosly reads from the c.incoming data channel and writes to the
+// current data buffer.
+func (c *Conn) readLoop() error {
+	for data := range c.incomingData {
+		c.currDataMut.Lock()
+		_, err := c.currData.Write(data)
+		if err != nil {
+			c.currDataMut.Unlock()
+			return err
+		}
+		c.currDataMut.Unlock()
+		c.dataSignal <- struct{}{}
```

Comment:
This will block until the data is read. It should be a channel-based flag (size-one buffer, write to it with a select).